packages/firestore/src/local/local_store.ts

```diff
@@ -642,6 +615,191 @@ export class LocalStore {
+        const existingDoc = existingDocs.get(key);
+        let docReadVersion: SnapshotVersion | null = null;
+        if (universalReadTime) {
+          docReadVersion = remoteVersion as SnapshotVersion;
+        } else {
+          docReadVersion = (remoteVersion as DocumentVersionMap).get(key);
+        }
+        debugAssert(!!docReadVersion, 'Document read version must exist');
+
+        // Note: The order of the steps below is important, since we want
+        // to ensure that rejected limbo resolutions (which fabricate
+        // NoDocuments with SnapshotVersion.min()) never add documents to
+        // cache.
+        if (
+          doc instanceof NoDocument &&
+          doc.version.isEqual(SnapshotVersion.min())
+        ) {
+          // NoDocuments with SnapshotVersion.min() are used in manufactured
+          // events. We remove these documents from cache since we lost
+          // access.
+          documentBuffer.removeEntry(key, docReadVersion!);
+          changedDocs = changedDocs.insert(key, doc);
+        } else if (
+          existingDoc == null ||
+          doc.version.compareTo(existingDoc.version) > 0 ||
+          (doc.version.compareTo(existingDoc.version) === 0 &&
+            existingDoc.hasPendingWrites)
+        ) {
+          debugAssert(
+            !SnapshotVersion.min().isEqual(docReadVersion!),
+            'Cannot add a document when the remote version is zero'
+          );
+          documentBuffer.addEntry(doc, docReadVersion!);
+          changedDocs = changedDocs.insert(key, doc);
+        } else {
+          logDebug(
+            LOG_TAG,
+            'Ignoring outdated watch update for ',
+            key,
+            '. Current version:',
+            existingDoc.version,
+            ' Watch version:',
+            doc.version
+          );
+        }
+      });
+      return changedDocs;
+    });
+  }
+
+  /**
+   * Returns a promise of a boolean to indicate if the given bundle has already
+   * been loaded and the create time is newer than the current loading bundle.
+   */
+  isNewerBundleLoaded(
+    bundleMetadata: bundleProto.BundleMetadata
+  ): Promise<boolean> {
+    const currentReadTime = this.bundleConverter.toSnapshotVersion(
+      bundleMetadata.createTime!
+    );
+    return this.persistence
+      .runTransaction('isNewerBundleLoaded', 'readonly', transaction => {
+        return this.bundleCache.getBundle(transaction, bundleMetadata.id!);
+      })
+      .then(cached => {
+        return !!cached && cached!.createTime!.compareTo(currentReadTime) > 0;
+      });
+  }
+
+  /**
+   * Saves the given `BundleMetadata` to local persistence.
+   * @param bundleMetadata
+   */
+  saveBundle(bundleMetadata: bundleProto.BundleMetadata): Promise<void> {
+    return this.persistence.runTransaction(
+      'Save bundle',
+      'readwrite',
+      transaction => {
+        return this.bundleCache.saveBundleMetadata(transaction, bundleMetadata);
```

Comment:
Should we also update `getBundle` to be `getBundleMetdadata`?. I just had to look up what `getBundle` does (to verify that it doesn't look a bunch of data in `isNewerBundleLoaded`).

Reply:
Yeah, I was trying to avoid using the same type name there. Changed to use `getBundleMetadata` now.